File: lib/csharp/src/structure/BasePropertyDeclarationSyntax.cpp
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2017 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

#include "csharp/inc/csharp.h"
#include "csharp/inc/Common.h"
#include "common/inc/WriteMessage.h"

#include "csharp/inc/messages.h"
#include <algorithm>
#include <string.h>
#include "common/inc/math/common.h"


namespace columbus { namespace csharp { namespace asg {

typedef boost::crc_32_type  Crc_type;

namespace structure { 
  BasePropertyDeclarationSyntax::BasePropertyDeclarationSyntax(NodeId _id, Factory *_factory) :
    MemberDeclarationSyntax(_id, _factory),
    m_AccessorList(0),
    AttributeListsContainer(),
    m_ExplicitInterfaceSpecifier(0),
    m_Type(0)
  {
  }

  BasePropertyDeclarationSyntax::~BasePropertyDeclarationSyntax() {
  }

  void BasePropertyDeclarationSyntax::prepareDelete(bool tryOnVirtualParent){
    removeAccessorList();
    while (!AttributeListsContainer.empty()) {
      const NodeId id = *AttributeListsContainer.begin();
      removeParentEdge(id);
      if (factory->getExistsReverseEdges())
        factory->reverseEdges->removeEdge(id, this->getId(), edkBasePropertyDeclarationSyntax_AttributeLists);
      AttributeListsContainer.pop_front();
    }
    removeExplicitInterfaceSpecifier();
    removeType();
    structure::MemberDeclarationSyntax::prepareDelete(false);
  }

  structure::AccessorListSyntax* BasePropertyDeclarationSyntax::getAccessorList() const {
    structure::AccessorListSyntax *_node = NULL;
    if (m_AccessorList != 0)
      _node = dynamic_cast<structure::AccessorListSyntax*>(factory->getPointer(m_AccessorList));
    if ( (_node == NULL) || factory->getIsFiltered(_node))
      return NULL;

    return _node;
  }

  ListIterator<structure::AttributeListSyntax> BasePropertyDeclarationSyntax::getAttributeListsListIteratorBegin() const {
    return ListIterator<structure::AttributeListSyntax>(&AttributeListsContainer, factory, true);
  }

  ListIterator<structure::AttributeListSyntax> BasePropertyDeclarationSyntax::getAttributeListsListIteratorEnd() const {
    return ListIterator<structure::AttributeListSyntax>(&AttributeListsContainer, factory, false);
  }

  bool BasePropertyDeclarationSyntax::getAttributeListsIsEmpty() const {
    return getAttributeListsListIteratorBegin() == getAttributeListsListIteratorEnd();
  }

  unsigned int BasePropertyDeclarationSyntax::getAttributeListsSize() const {
    unsigned int size = 0;
    ListIterator<structure::AttributeListSyntax> endIt = getAttributeListsListIteratorEnd();
    for (ListIterator<structure::AttributeListSyntax> it = getAttributeListsListIteratorBegin(); it != endIt; ++it) {
      ++size;
    }
    return size;
  }

  structure::ExplicitInterfaceSpecifierSyntax* BasePropertyDeclarationSyntax::getExplicitInterfaceSpecifier() const {
    structure::ExplicitInterfaceSpecifierSyntax *_node = NULL;
    if (m_ExplicitInterfaceSpecifier != 0)
      _node = dynamic_cast<structure::ExplicitInterfaceSpecifierSyntax*>(factory->getPointer(m_ExplicitInterfaceSpecifier));
    if ( (_node == NULL) || factory->getIsFiltered(_node))
      return NULL;

    return _node;
  }

  expression::TypeSyntax* BasePropertyDeclarationSyntax::getType() const {
    expression::TypeSyntax *_node = NULL;
    if (m_Type != 0)
      _node = dynamic_cast<expression::TypeSyntax*>(factory->getPointer(m_Type));
    if ( (_node == NULL) || factory->getIsFiltered(_node))
      return NULL;

    return _node;
  }

  bool BasePropertyDeclarationSyntax::setEdge(EdgeKind edgeKind, NodeId edgeEnd, bool tryOnVirtualParent) {
    switch (edgeKind) {
      case edkBasePropertyDeclarationSyntax_AccessorList:
        setAccessorList(edgeEnd);
        return true;
      case edkBasePropertyDeclarationSyntax_AttributeLists:
        addAttributeLists(edgeEnd);
        return true;
      case edkBasePropertyDeclarationSyntax_ExplicitInterfaceSpecifier:
        setExplicitInterfaceSpecifier(edgeEnd);
        return true;
      case edkBasePropertyDeclarationSyntax_Type:
        setType(edgeEnd);
        return true;
      default:
        break;
    }
    if (structure::MemberDeclarationSyntax::setEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    return false;
  }

  bool BasePropertyDeclarationSyntax::removeEdge(EdgeKind edgeKind, NodeId edgeEnd, bool tryOnVirtualParent) {
    switch (edgeKind) {
      case edkBasePropertyDeclarationSyntax_AccessorList:
        removeAccessorList();
        return true;
      case edkBasePropertyDeclarationSyntax_AttributeLists:
        removeAttributeLists(edgeEnd);
        return true;
      case edkBasePropertyDeclarationSyntax_ExplicitInterfaceSpecifier:
        removeExplicitInterfaceSpecifier();
        return true;
      case edkBasePropertyDeclarationSyntax_Type:
        removeType();
        return true;
      default:
        break;
    }
    if (structure::MemberDeclarationSyntax::removeEdge(edgeKind, edgeEnd, false)) {
      return true;
    }
    return false;
  }

  void BasePropertyDeclarationSyntax::setAccessorList(NodeId _id) {
    structure::AccessorListSyntax *_node = NULL;
    if (_id) {
      if (!factory->getExist(_id))
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_END_POINT_OF_THE_EDGE_DOES_NOT_EXIST);

      _node = dynamic_cast<structure::AccessorListSyntax*> (factory->getPointer(_id));
      if ( _node == NULL) {
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_INVALID_NODE_KIND);
      }
      if (&(_node->getFactory()) != this->factory)
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_FACTORY_OF_NODES_DOES_NOT_MATCH );

      if (m_AccessorList) {
        removeParentEdge(m_AccessorList);
        if (factory->getExistsReverseEdges())
          factory->reverseEdges->removeEdge(m_AccessorList, m_id, edkBasePropertyDeclarationSyntax_AccessorList);
      }
      m_AccessorList = _node->getId();
      if (m_AccessorList != 0)
        setParentEdge(factory->getPointer(m_AccessorList), edkBasePropertyDeclarationSyntax_AccessorList);
      if (factory->getExistsReverseEdges())
        factory->reverseEdges->insertEdge(m_AccessorList, this->getId(), edkBasePropertyDeclarationSyntax_AccessorList);
    } else {
      if (m_AccessorList) {
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_CAN_T_SET_EDGE_TO_NULL);
      }
    }
  }

  void BasePropertyDeclarationSyntax::setAccessorList(structure::AccessorListSyntax *_node) {
    if (_node == NULL)
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_CAN_T_SET_EDGE_TO_NULL);

    setAccessorList(_node->getId());
  }

  void BasePropertyDeclarationSyntax::removeAccessorList() {
      if (m_AccessorList) {
        removeParentEdge(m_AccessorList);
        if (factory->getExistsReverseEdges())
          factory->reverseEdges->removeEdge(m_AccessorList, m_id, edkBasePropertyDeclarationSyntax_AccessorList);
      }
      m_AccessorList = 0;
  }

  void BasePropertyDeclarationSyntax::addAttributeLists(const structure::AttributeListSyntax *_node) {
    if (_node == NULL)
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_NODE_IS_NULL);

    if (&(_node->getFactory()) != this->factory)
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_FACTORY_OF_NODES_DOES_NOT_MATCH);

    if (!((_node->getNodeKind() == ndkAttributeListSyntax) ))
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_INVALID_NODE_KIND);

    AttributeListsContainer.push_back(_node->getId());
    setParentEdge(_node,edkBasePropertyDeclarationSyntax_AttributeLists);

    if (factory->reverseEdges)
      factory->reverseEdges->insertEdge(_node, this, edkBasePropertyDeclarationSyntax_AttributeLists);
  }

  void BasePropertyDeclarationSyntax::addAttributeLists(NodeId _id) {
    const structure::AttributeListSyntax *node = dynamic_cast<structure::AttributeListSyntax*>(factory->getPointer(_id));
    if (node == NULL)
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_INVALID_NODE_KIND);
    addAttributeLists( node );
  }

  void BasePropertyDeclarationSyntax::removeAttributeLists(NodeId id) {
    if (!factory->getExist(id))
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_END_POINT_OF_THE_EDGE_DOES_NOT_EXIST);

    ListIterator<structure::AttributeListSyntax>::Container::iterator it = find(AttributeListsContainer.begin(), AttributeListsContainer.end(), id);

    if (it == AttributeListsContainer.end())
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_END_POINT_OF_THE_EDGE_DOES_NOT_EXIST);

    AttributeListsContainer.erase(it);

    removeParentEdge(id);

    if (factory->getExistsReverseEdges())
      factory->reverseEdges->removeEdge(id, this->getId(), edkBasePropertyDeclarationSyntax_AttributeLists);
  }

  void BasePropertyDeclarationSyntax::removeAttributeLists(structure::AttributeListSyntax *_node) {
    if (_node == NULL)
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_EDGE_IS_NULL);

    removeAttributeLists(_node->getId());
  }

  void BasePropertyDeclarationSyntax::setExplicitInterfaceSpecifier(NodeId _id) {
    structure::ExplicitInterfaceSpecifierSyntax *_node = NULL;
    if (_id) {
      if (!factory->getExist(_id))
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_END_POINT_OF_THE_EDGE_DOES_NOT_EXIST);

      _node = dynamic_cast<structure::ExplicitInterfaceSpecifierSyntax*> (factory->getPointer(_id));
      if ( _node == NULL) {
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_INVALID_NODE_KIND);
      }
      if (&(_node->getFactory()) != this->factory)
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_FACTORY_OF_NODES_DOES_NOT_MATCH );

      if (m_ExplicitInterfaceSpecifier) {
        removeParentEdge(m_ExplicitInterfaceSpecifier);
        if (factory->getExistsReverseEdges())
          factory->reverseEdges->removeEdge(m_ExplicitInterfaceSpecifier, m_id, edkBasePropertyDeclarationSyntax_ExplicitInterfaceSpecifier);
      }
      m_ExplicitInterfaceSpecifier = _node->getId();
      if (m_ExplicitInterfaceSpecifier != 0)
        setParentEdge(factory->getPointer(m_ExplicitInterfaceSpecifier), edkBasePropertyDeclarationSyntax_ExplicitInterfaceSpecifier);
      if (factory->getExistsReverseEdges())
        factory->reverseEdges->insertEdge(m_ExplicitInterfaceSpecifier, this->getId(), edkBasePropertyDeclarationSyntax_ExplicitInterfaceSpecifier);
    } else {
      if (m_ExplicitInterfaceSpecifier) {
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_CAN_T_SET_EDGE_TO_NULL);
      }
    }
  }

  void BasePropertyDeclarationSyntax::setExplicitInterfaceSpecifier(structure::ExplicitInterfaceSpecifierSyntax *_node) {
    if (_node == NULL)
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_CAN_T_SET_EDGE_TO_NULL);

    setExplicitInterfaceSpecifier(_node->getId());
  }

  void BasePropertyDeclarationSyntax::removeExplicitInterfaceSpecifier() {
      if (m_ExplicitInterfaceSpecifier) {
        removeParentEdge(m_ExplicitInterfaceSpecifier);
        if (factory->getExistsReverseEdges())
          factory->reverseEdges->removeEdge(m_ExplicitInterfaceSpecifier, m_id, edkBasePropertyDeclarationSyntax_ExplicitInterfaceSpecifier);
      }
      m_ExplicitInterfaceSpecifier = 0;
  }

  void BasePropertyDeclarationSyntax::setType(NodeId _id) {
    expression::TypeSyntax *_node = NULL;
    if (_id) {
      if (!factory->getExist(_id))
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_END_POINT_OF_THE_EDGE_DOES_NOT_EXIST);

      _node = dynamic_cast<expression::TypeSyntax*> (factory->getPointer(_id));
      if ( _node == NULL) {
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_INVALID_NODE_KIND);
      }
      if (&(_node->getFactory()) != this->factory)
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_THE_FACTORY_OF_NODES_DOES_NOT_MATCH );

      if (m_Type) {
        removeParentEdge(m_Type);
        if (factory->getExistsReverseEdges())
          factory->reverseEdges->removeEdge(m_Type, m_id, edkBasePropertyDeclarationSyntax_Type);
      }
      m_Type = _node->getId();
      if (m_Type != 0)
        setParentEdge(factory->getPointer(m_Type), edkBasePropertyDeclarationSyntax_Type);
      if (factory->getExistsReverseEdges())
        factory->reverseEdges->insertEdge(m_Type, this->getId(), edkBasePropertyDeclarationSyntax_Type);
    } else {
      if (m_Type) {
        throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_CAN_T_SET_EDGE_TO_NULL);
      }
    }
  }

  void BasePropertyDeclarationSyntax::setType(expression::TypeSyntax *_node) {
    if (_node == NULL)
      throw CsharpException(COLUMBUS_LOCATION, CMSG_EX_CAN_T_SET_EDGE_TO_NULL);

    setType(_node->getId());
  }

  void BasePropertyDeclarationSyntax::removeType() {
      if (m_Type) {
        removeParentEdge(m_Type);
        if (factory->getExistsReverseEdges())
          factory->reverseEdges->removeEdge(m_Type, m_id, edkBasePropertyDeclarationSyntax_Type);
      }
      m_Type = 0;
  }

  double BasePropertyDeclarationSyntax::getSimilarity(const base::Base& base){
    if(base.getNodeKind() == getNodeKind()) {
      return 1.0;
    } else {
      return 0.0;
    }
  }

  void BasePropertyDeclarationSyntax::swapStringTable(RefDistributorStrTable& newStrTable, std::map<Key,Key>& oldAndNewStrKeyMap ){
    std::map<Key,Key>::iterator foundKeyId;

      m_position.swapStringTable(newStrTable,oldAndNewStrKeyMap);

  }

  NodeHashType BasePropertyDeclarationSyntax::getHash(std::set<NodeId>& travNodes) const {
    if (hashOk) return nodeHashCache;
    common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_BEGIN,this->getId()); 
    if (travNodes.count(getId())>0) {
      common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_SKIP);
      return 0;
    }
    travNodes.insert(getId());
    Crc_type resultHash;
    resultHash.process_bytes( "structure::BasePropertyDeclarationSyntax", strlen("structure::BasePropertyDeclarationSyntax"));
    common::WriteMsg::write(CMSG_GET_THE_NODE_HASH_OF_NODE_END,resultHash.checksum()); 
    nodeHashCache = resultHash.checksum();
    hashOk = true;
    return nodeHashCache;
  }

  void BasePropertyDeclarationSyntax::save(io::BinaryIO &binIo,bool withVirtualBase  /*= true*/) const {
    MemberDeclarationSyntax::save(binIo,false);

    binIo.writeUInt4(m_AccessorList);
    binIo.writeUInt4(m_ExplicitInterfaceSpecifier);
    binIo.writeUInt4(m_Type);


    for (ListIterator<structure::AttributeListSyntax>::Container::const_iterator it = AttributeListsContainer.begin(); it != AttributeListsContainer.end(); ++it) {
      binIo.writeUInt4(*it);
    }
    binIo.writeUInt4(0); // This is the end sign
  }

  void BasePropertyDeclarationSyntax::load(io::BinaryIO &binIo, bool withVirtualBase /*= true*/) {
    MemberDeclarationSyntax::load(binIo,false);

    m_AccessorList =  binIo.readUInt4();
    if (m_AccessorList != 0)
      setParentEdge(factory->getPointer(m_AccessorList),edkBasePropertyDeclarationSyntax_AccessorList);

    m_ExplicitInterfaceSpecifier =  binIo.readUInt4();
    if (m_ExplicitInterfaceSpecifier != 0)
      setParentEdge(factory->getPointer(m_ExplicitInterfaceSpecifier),edkBasePropertyDeclarationSyntax_ExplicitInterfaceSpecifier);

    m_Type =  binIo.readUInt4();
    if (m_Type != 0)
      setParentEdge(factory->getPointer(m_Type),edkBasePropertyDeclarationSyntax_Type);

    NodeId _id;

    _id = binIo.readUInt4();
    while (_id) {
      AttributeListsContainer.push_back(_id);
      setParentEdge(factory->getPointer(_id),edkBasePropertyDeclarationSyntax_AttributeLists);
      _id = binIo.readUInt4();
    }
  }


}


}}}

File: csharp/lib/csharp/Nodes/Structure/BasePropertyDeclarationSyntax.cs
/*
 *  This file is part of OpenStaticAnalyzer.
 *
 *  Copyright (c) 2004-2018 Department of Software Engineering - University of Szeged
 *
 *  Licensed under Version 1.2 of the EUPL (the "Licence");
 *
 *  You may not use this work except in compliance with the Licence.
 *
 *  You may obtain a copy of the Licence in the LICENSE file or at:
 *
 *  https://joinup.ec.europa.eu/software/page/eupl
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the Licence is distributed on an "AS IS" basis,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the Licence for the specific language governing permissions and
 *  limitations under the Licence.
 */

using System;
using System.Text;
using System.Diagnostics;
using System.Collections.Generic;

namespace Columbus.Csharp.Asg.Nodes.Structure {

    /// <summary>
    /// BasePropertyDeclarationSyntax class, which represents the structure::BasePropertyDeclarationSyntax node.
    /// </summary>
    [DebuggerDisplay("{DemangledName}")]
#if (CSHARP_INTERNAL)
    internal
#else
    public
#endif
    class BasePropertyDeclarationSyntax : MemberDeclarationSyntax {


        // ---------- Edges ----------

        /// <summary>The id of the node the AccessorList edge points to.</summary>
        protected uint m_AccessorList;

        /// <summary>Container stores the id of the nodes the AttributeLists edge points to.</summary>
        protected LinkedList<uint> AttributeListsContainer;

        /// <summary>The id of the node the ExplicitInterfaceSpecifier edge points to.</summary>
        protected uint m_ExplicitInterfaceSpecifier;

        /// <summary>The id of the node the Type edge points to.</summary>
        protected uint m_Type;

        /// <summary>
        /// Constructor, only factory can instantiates nodes.
        /// </summary>
        /// <param name="nodeId">[in] The id of the node.</param>
        /// <param name="factory">[in] Poiter to the Factory the node belongs to.</param>
        public BasePropertyDeclarationSyntax(uint nodeId, Factory factory) : base(nodeId, factory) {
            m_AccessorList = 0;
            AttributeListsContainer = new LinkedList<uint>();
            m_ExplicitInterfaceSpecifier = 0;
            m_Type = 0;
        }

        /// <summary>
        /// Gives back the NodeKind of the node.
        /// </summary>
        /// <returns>Returns with the NodeKind.</returns>
        public override Types.NodeKind NodeKind {
            get { return Types.NodeKind.ndkBasePropertyDeclarationSyntax; }
        }


        // ---------- Reflection getter function ----------

        /// <summary>
        /// Gives back the begin iterator for the edge specified
        /// </summary>
        /// <param name="edge">[in] The multiple edge for which the iterator is requested</param>
        public override ListIterator<Columbus.Csharp.Asg.Nodes.Base.Base> GetListIteratorBegin(string edge) {
            switch(edge) {
                case "AttributeLists":
                    return AttributeListsListIteratorBegin.Convert<Columbus.Csharp.Asg.Nodes.Base.Base>();
                default:
                    return base.GetListIteratorBegin(edge);
            }
        }


        // ---------- Reflection add function ----------

        /// <summary>
        /// Adds item as a new edge
        /// </summary>
        public override void add(string edge, uint nodeId) {
            switch(edge) {
                case "AttributeLists":
                    addAttributeLists(nodeId);
                    return;
                default:
                    base.add(edge, nodeId);
                    return;
            }
        }


        // ---------- Edge getter function(s) ----------

        /// <summary>
        /// Gives back the id of the node the AccessorList edge points to.
        /// </summary>
        /// <returns>Returns the end point of the m_AccessorList edge.</returns>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if the edge id is invalid.</exception>
        public uint getAccessorList() {
            if (fact.getIsFiltered(m_AccessorList))
                return 1;
            else
                return m_AccessorList;
        }

        /// <summary>
        /// Gives back iterator for the AttributeLists edges.
        /// </summary>
        /// <returns>Returns an iterator for the AttributeLists edges.</returns>
        public ListIterator<Columbus.Csharp.Asg.Nodes.Structure.AttributeListSyntax> AttributeListsListIteratorBegin {
            get {
                return new ListIterator<Columbus.Csharp.Asg.Nodes.Structure.AttributeListSyntax>(AttributeListsContainer, fact, true);
            }
        }

        /// <summary>
        /// Gives back iterator for the AttributeLists edges.
        /// </summary>
        /// <returns>Returns an iterator for the AttributeLists edges.</returns>
        public ListIterator<Columbus.Csharp.Asg.Nodes.Structure.AttributeListSyntax> AttributeListsListIteratorEnd {
            get {
                return new ListIterator<Columbus.Csharp.Asg.Nodes.Structure.AttributeListSyntax>(AttributeListsContainer, fact, false);
            }
        }

        /// <summary>
        /// Tells whether the node has AttributeLists edges or not.
        /// </summary>
        /// <returns>Returns true if the node has at least one AttributeLists edge.</returns>
        public bool AttributeListsIsEmpty {
            get { return AttributeListsContainer.Count == 0; }
        }

        /// <summary>
        /// Gives back how many AttributeLists edges the node has.
        /// </summary>
        /// <returns>Returns with the number of AttributeLists edges.</returns>
        public uint AttributeListsSize {
            get { return (uint)(AttributeListsContainer.Count); }
        }

        /// <summary>
        /// Gives back the id of the node the ExplicitInterfaceSpecifier edge points to.
        /// </summary>
        /// <returns>Returns the end point of the m_ExplicitInterfaceSpecifier edge.</returns>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if the edge id is invalid.</exception>
        public uint getExplicitInterfaceSpecifier() {
            if (fact.getIsFiltered(m_ExplicitInterfaceSpecifier))
                return 1;
            else
                return m_ExplicitInterfaceSpecifier;
        }

        /// <summary>
        /// Gives back the id of the node the Type edge points to.
        /// </summary>
        /// <returns>Returns the end point of the m_Type edge.</returns>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if the edge id is invalid.</exception>
        public uint getType() {
            if (fact.getIsFiltered(m_Type))
                return 1;
            else
                return m_Type;
        }


        // ---------- Edge setter function(s) ----------

        /// <summary>
        /// Sets the AccessorList edge.
        /// </summary>
        /// <param name="nodeId">[in] The new end point of the AccessorList edge or 0. With 0 the edge can be deleted.</param>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if the nodeId is invalid.</exception>
        public void setAccessorList(uint nodeId) {
            if (nodeId != 0) {
                if (!fact.getExist(nodeId))
                    throw new CsharpException("Columbus.Csharp.Asg.Nodes.Structure.BasePropertyDeclarationSyntax.setAccessorList(NodeId)", "The end point of the edge does not exist.");

                Columbus.Csharp.Asg.Nodes.Base.Base node = fact.getRef(nodeId);
                if (Common.getIsBaseClassKind(node.NodeKind, Types.NodeKind.ndkAccessorListSyntax)) {
                    if (m_AccessorList != 0) {
                        removeParentEdge(m_AccessorList);
                    }
                    m_AccessorList = nodeId;
                    setParentEdge(m_AccessorList, (uint)Types.EdgeKind.edkBasePropertyDeclarationSyntax_AccessorList);
                } else {
                    throw new CsharpException("Columbus.Csharp.Asg.Nodes.Structure.BasePropertyDeclarationSyntax.setAccessorList(NodeId)", "Invalid NodeKind (" + node.NodeKind.ToString() + ").");
                }
            } else {
                if (m_AccessorList != 0) {
                    removeParentEdge(m_AccessorList);
                }
                m_AccessorList = 0;
            }
        }

        /// <summary>
        /// Sets the AccessorList edge.
        /// </summary>
        /// <param name="node">[in] The new end point of the AccessorList edge.</param>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if there's something wrong with the given node.</exception>
        public void setAccessorList(Columbus.Csharp.Asg.Nodes.Structure.AccessorListSyntax node) {
            if (m_AccessorList != 0) {
                removeParentEdge(m_AccessorList);
            }
            m_AccessorList = node.Id;
            setParentEdge(m_AccessorList, (uint)Types.EdgeKind.edkBasePropertyDeclarationSyntax_AccessorList);
        }

        /// <summary>
        /// Adds a new AttributeLists edge to the node and inserts it after the other ones.
        /// </summary>
        /// <param name="node">[in] The end point of the new AttributeLists edge.</param>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if there is something wrong with the given node.</exception>
        public void addAttributeLists(Columbus.Csharp.Asg.Nodes.Structure.AttributeListSyntax node) {
            addAttributeLists(node.Id);
        }

        /// <summary>
        /// Adds a new AttributeLists edge to the node and inserts it after the other ones.
        /// </summary>
        /// <param name="nodeId">[in] The end point of the new AttributeLists edge.</param>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if the given nodeId is invalid.</exception>
        public void addAttributeLists(uint nodeId) {
            if (!fact.getExist(nodeId))
                throw new CsharpException("Columbus.Csharp.Asg.Nodes.Structure.BasePropertyDeclarationSyntax.addAttributeLists(NodeId)", "The end point of the edge does not exist.");
            Columbus.Csharp.Asg.Nodes.Base.Base node = fact.getRef(nodeId);
            if (Common.getIsBaseClassKind(node.NodeKind, Types.NodeKind.ndkAttributeListSyntax)) {
                AttributeListsContainer.AddLast(nodeId);
            } else {
                throw new CsharpException("Columbus.Csharp.Asg.Nodes.Structure.BasePropertyDeclarationSyntax.setAttributeLists(NodeId)", "Invalid NodeKind (" + node.NodeKind.ToString() + ").");
            }
           setParentEdge(nodeId, (uint)Types.EdgeKind.edkBasePropertyDeclarationSyntax_AttributeLists);
            if (fact.ReverseEdges != null)
              fact.ReverseEdges.insertEdge(nodeId, this.Id, Types.EdgeKind.edkBasePropertyDeclarationSyntax_AttributeLists);
        }

        /// <summary>
        /// Sets the ExplicitInterfaceSpecifier edge.
        /// </summary>
        /// <param name="nodeId">[in] The new end point of the ExplicitInterfaceSpecifier edge or 0. With 0 the edge can be deleted.</param>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if the nodeId is invalid.</exception>
        public void setExplicitInterfaceSpecifier(uint nodeId) {
            if (nodeId != 0) {
                if (!fact.getExist(nodeId))
                    throw new CsharpException("Columbus.Csharp.Asg.Nodes.Structure.BasePropertyDeclarationSyntax.setExplicitInterfaceSpecifier(NodeId)", "The end point of the edge does not exist.");

                Columbus.Csharp.Asg.Nodes.Base.Base node = fact.getRef(nodeId);
                if (Common.getIsBaseClassKind(node.NodeKind, Types.NodeKind.ndkExplicitInterfaceSpecifierSyntax)) {
                    if (m_ExplicitInterfaceSpecifier != 0) {
                        removeParentEdge(m_ExplicitInterfaceSpecifier);
                    }
                    m_ExplicitInterfaceSpecifier = nodeId;
                    setParentEdge(m_ExplicitInterfaceSpecifier, (uint)Types.EdgeKind.edkBasePropertyDeclarationSyntax_ExplicitInterfaceSpecifier);
                } else {
                    throw new CsharpException("Columbus.Csharp.Asg.Nodes.Structure.BasePropertyDeclarationSyntax.setExplicitInterfaceSpecifier(NodeId)", "Invalid NodeKind (" + node.NodeKind.ToString() + ").");
                }
            } else {
                if (m_ExplicitInterfaceSpecifier != 0) {
                    removeParentEdge(m_ExplicitInterfaceSpecifier);
                }
                m_ExplicitInterfaceSpecifier = 0;
            }
        }

        /// <summary>
        /// Sets the ExplicitInterfaceSpecifier edge.
        /// </summary>
        /// <param name="node">[in] The new end point of the ExplicitInterfaceSpecifier edge.</param>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if there's something wrong with the given node.</exception>
        public void setExplicitInterfaceSpecifier(Columbus.Csharp.Asg.Nodes.Structure.ExplicitInterfaceSpecifierSyntax node) {
            if (m_ExplicitInterfaceSpecifier != 0) {
                removeParentEdge(m_ExplicitInterfaceSpecifier);
            }
            m_ExplicitInterfaceSpecifier = node.Id;
            setParentEdge(m_ExplicitInterfaceSpecifier, (uint)Types.EdgeKind.edkBasePropertyDeclarationSyntax_ExplicitInterfaceSpecifier);
        }

        /// <summary>
        /// Sets the Type edge.
        /// </summary>
        /// <param name="nodeId">[in] The new end point of the Type edge or 0. With 0 the edge can be deleted.</param>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if the nodeId is invalid.</exception>
        public void setType(uint nodeId) {
            if (nodeId != 0) {
                if (!fact.getExist(nodeId))
                    throw new CsharpException("Columbus.Csharp.Asg.Nodes.Structure.BasePropertyDeclarationSyntax.setType(NodeId)", "The end point of the edge does not exist.");

                Columbus.Csharp.Asg.Nodes.Base.Base node = fact.getRef(nodeId);
                if (Common.getIsBaseClassKind(node.NodeKind, Types.NodeKind.ndkTypeSyntax)) {
                    if (m_Type != 0) {
                        removeParentEdge(m_Type);
                    }
                    m_Type = nodeId;
                    setParentEdge(m_Type, (uint)Types.EdgeKind.edkBasePropertyDeclarationSyntax_Type);
                } else {
                    throw new CsharpException("Columbus.Csharp.Asg.Nodes.Structure.BasePropertyDeclarationSyntax.setType(NodeId)", "Invalid NodeKind (" + node.NodeKind.ToString() + ").");
                }
            } else {
                if (m_Type != 0) {
                    removeParentEdge(m_Type);
                }
                m_Type = 0;
            }
        }

        /// <summary>
        /// Sets the Type edge.
        /// </summary>
        /// <param name="node">[in] The new end point of the Type edge.</param>
        /// <exception cref="Columbus.Csharp.Asg.CsharpException">Throws CsharpException if there's something wrong with the given node.</exception>
        public void setType(Columbus.Csharp.Asg.Nodes.Expression.TypeSyntax node) {
            if (m_Type != 0) {
                removeParentEdge(m_Type);
            }
            m_Type = node.Id;
            setParentEdge(m_Type, (uint)Types.EdgeKind.edkBasePropertyDeclarationSyntax_Type);
        }

        /// <summary>
        /// Saves the node.
        /// </summary>
        /// <param name="io">[in] The node is written into io.</param>
        /// <exception cref="Columbus.ColumbusIOException">Throws ColumbusIOException if there is something wrong with the file.</exception>
        public override void save(IO io) {
            base.save(io);

            io.writeUInt4(m_AccessorList);
            io.writeUInt4(m_ExplicitInterfaceSpecifier);
            io.writeUInt4(m_Type);


            foreach (var AttributeLists in AttributeListsContainer) {
              io.writeUInt4(AttributeLists);
            }
            io.writeUInt4(0); // This is the end sign
        }

        /// <summary>
        /// Loads the node.
        /// </summary>
        /// <param name="io">[in] The node is read from io.</param>
        /// <exception cref="Columbus.ColumbusIOException">Throws ColumbusIOException if there is something wrong with the file.</exception>
        public override void load(IO binIo) {
            base.load(binIo);

            m_AccessorList =  binIo.readUInt4();
            if (m_AccessorList != 0)
              setParentEdge(m_AccessorList,(uint)Types.EdgeKind.edkBasePropertyDeclarationSyntax_AccessorList);

            m_ExplicitInterfaceSpecifier =  binIo.readUInt4();
            if (m_ExplicitInterfaceSpecifier != 0)
              setParentEdge(m_ExplicitInterfaceSpecifier,(uint)Types.EdgeKind.edkBasePropertyDeclarationSyntax_ExplicitInterfaceSpecifier);

            m_Type =  binIo.readUInt4();
            if (m_Type != 0)
              setParentEdge(m_Type,(uint)Types.EdgeKind.edkBasePropertyDeclarationSyntax_Type);

            uint _id;

            _id = binIo.readUInt4();
            while (_id != 0) {
              AttributeListsContainer.AddLast(_id);
              setParentEdge(_id, (uint)Types.EdgeKind.edkBasePropertyDeclarationSyntax_AttributeLists);
              _id = binIo.readUInt4();
            }
        }

    }

}
